utils/plugin/plugin_downloader.go

```diff
@@ -0,0 +1,190 @@
+)
+
+// FileDownloader is an interface that allows us to mock the http.Get function
+type FileDownloader interface {
+	Get(url string) (resp *http.Response, err error)
+}
+
+// FileDownloaderImpl is the default/real implementation of the FileDownloader interface
+type FileDownloaderImpl struct {
+}
+
+func (fd FileDownloaderImpl) Get(url string) (resp *http.Response, err error) {
+	return http.Get(url)
+}
+
+// checkPluginExists this function checks if the plugin exists in the configured path if not we panic
+func checkPluginExists(pluginLocation string) error {
+	if pluginLocation != "" {
+		//Check for plugin executable existence
+		_, err := os.Stat(pluginLocation)
+		if err != nil {
+			return fmt.Errorf("plugin stat file at %s", pluginLocation)
+		}
+	}
+	return nil
+}
+
+func checkShaOfPlugin(pluginLocation string, expectedSha256 string) (bool, error) {
+	hasher := sha256.New()
+	fileBytes, err := os.ReadFile(pluginLocation)
+	if err != nil {
+		return false, fmt.Errorf("failed to read file %s: %w", pluginLocation, err)
+	}
+	fileSha256 := fmt.Sprintf("%x", hasher.Sum(fileBytes))
+	match := fileSha256 == expectedSha256
+	if !match {
+		log.Printf("expected sha256: %s, actual sha256: %s, of downloaded metric plugin (%s)", expectedSha256, fileSha256, pluginLocation)
+	}
+	return match, nil
+}
+
+func downloadFile(filepath string, url string, downloader FileDownloader) error {
+	// Get the data
+	resp, err := downloader.Get(url)
+	if err != nil {
+		return fmt.Errorf("failed to download file from %s: %w", url, err)
+	}
+	defer resp.Body.Close()
+
+	// Create the file
+	out, err := os.Create(filepath)
+	if err != nil {
+		return fmt.Errorf("failed to create file %s: %w", filepath, err)
+	}
+	defer out.Close()
+
+	// Write the body to file
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to write to file %s: %w", filepath, err)
+	}
+
+	// Set the file permissions, to allow execution
+	err = os.Chmod(filepath, 0700)
+	if err != nil {
+		return fmt.Errorf("failed to set file permissions on %s: %w", filepath, err)
+	}
+
+	return err
+}
+
+// DownloadPlugins this function downloads and/or checks that a plugin executable exits on the filesystem
+func DownloadPlugins(fd FileDownloader) error {
+	config, err := config.GetConfig()
+	if err != nil {
+		return fmt.Errorf("failed to get config: %w", err)
+	}
+
+	absoluteFilepath, err := filepath.Abs(defaults.DefaultRolloutPluginFolder)
+	if err != nil {
+		return fmt.Errorf("failed to get absolute path of plugin folder: %w", err)
+	}
+
+	err = os.MkdirAll(absoluteFilepath, 0700)
+	if err != nil {
+		return fmt.Errorf("failed to create plugin folder: %w", err)
+	}
+
+	for _, plugin := range config.GetMetricPluginsConfig() {
+		urlObj, err := url.ParseRequestURI(plugin.PluginLocation)
+		if err != nil {
+			return fmt.Errorf("failed to parse plugin location: %w", err)
+		}
+
+		finalFileLocation := filepath.Join(absoluteFilepath, plugin.Name)
```

Comment:
how are we going to handle 2 plugins with the same name? wouldn't it be better to have some sort of namespace for plugins as well?

Reply:
correct this does not support two plugins with the same name I am not sure of the value 100% with that and it is somewhat controlled via the plugin themself because it is the registration method as well, aka plugins need to know where to look for their config at. I think we could provide a best practice as a plugin author to namespace them. Do you have any thoughts on what would make sense for namespace?

Reply:
Talking with leo we will probably try to enforce a strict file naming convention based on git repo to allow multiple plugins.